### DISCUSSION TITLE
Ctrl-n on empty line steps forward from previous history fetch

### DISCUSSION THREAD

**Discussion by albheim** [fish-shell/fish-shell#10070] [Ideas]
I find myself often having a sequence of commands that is run for some task, work on some more things and realise I need to run the same sequence again.
What I usually do is just ctrl-p until i find the first, and remember how many I did, and just repeat the same amount for each command (given it was not too far).

What I found is that the Julia REPL allows you to step forward from last history fetch using ctrl-n on an empty line (probably exists elsewhere also, but his is where I first encountered it).
This enables me to only make a search for the first command in the sequence (either ctrl-p or crtl-r) and run it, and then ctrl-n on an empty line will start stepping forward from that spot in history.
So to repeat my sequence I just need to find and run the first command in the sequence, and then do ctrl-n + enter a few times.

Running ctrl-n when the previous command was not fetched from history does nothing.

I don't know if this already exists under some other key-combination or command, or if it would be easy to build as a script, but I couldn't find it or figure it out based on what seemed available from the history command. Though I can't say I have tried very hard, pretty new to fish scripting.

If anyone knows if this exists or can easily be built it would be awesome, otherwise I guess it is an idea for a new feature that at least one person would like :)

**Comment by faho** [comment]
ctrl-n is bound to "down-or-search", this will either move down a line or, if you are currently in a history search, continue the search *forward* in time, i.e. to the newer element.

However it doesn't *start* a history search since, conceptually, you're always "before" the newest element when you enter a new command (basically the current commandline is future history).

And once you've *selected* a commandline from history, the history search ends. There's no persistent "history pointer" that tells fish you're still at that position in history and continues a search from there. This is a bit of state that can easily become confusing.

What you're asking for is essentially #9751.

**Reply by albheim** [reply]
Thanks for the reply!

Yeah, I thought it might not exist any state based on the docs of the history command.

I'll have a look at that issue and see if I can make the suggestion it work for me, thanks for the link!

**Reply by albheim** [reply]
I don't think it is really what I want.

Is there any way of accessing the history based on an index? Can't find that in the documentation.

**Reply by faho** [reply]
$history contains the history from newest to oldest. $history[1] is the last command you ran.

**Reply by albheim** [reply]
Ahh, thanks! Tried `history[1]` but that didn't work (as expected, but I'm new and lost in fish scripting) and went down another path looking for the definition of `history_search_forward` in the c++ code... 
But with this I think I might be able to build a simple version of what I want.

**Reply by albheim** [reply]
If anyone else is interested I got it semi-working by copying and modding the `down-or-search/up-or-search` function.
It is good enough for me but not really all the way I would have wanted.
The following code will keep a global index tracking the last time a history item was fetched with ctrl-p/ctrl-n, and on ctrl-n on an empty line will jump back based on that index. It will also do this even when last command wasn't a fetch from history, don't know how I would avoid this. And it will not work on items fetched with ctrl-r or prefix searches with ctrl-p/ctrl-n either, though this could probably be fixed but I couldn't be bothered for now.

```
function my_hist_back --description 'Save history position'
    # If we are already in search mode, continue
    if commandline --search-mode
        commandline -f history-search-backward
        set -g LAST_FETCH_IDX (math $LAST_FETCH_IDX + 1)
        return
    end

    # If we are navigating the pager, then up always navigates
    if commandline --paging-mode
        commandline -f up-line
        return
    end

    # We are not already in search mode.
    # If we are on the top line, start search mode,
    # otherwise move up
    set -l lineno (commandline -L)

    switch $lineno
        case 1
            commandline -f history-search-backward
            set -g LAST_FETCH_IDX 1

        case '*'
            commandline -f up-line
    end
end

function my_hist_forward --description 'Navigate history'
    # If we are already in search mode, continue
    if commandline --search-mode
        commandline -f history-search-forward
        set -g LAST_FETCH_IDX (math $LAST_FETCH_IDX - 1)
        return
    end

    # If we are navigating the pager, then up always navigates
    if commandline --paging-mode
        commandline -f down-line
        return
    end

    if test -z (commandline) && test -n $LAST_FETCH
        commandline -f history-search-backward
        set target $history[$LAST_FETCH_IDX]
        for i in (seq $LAST_FETCH_IDX)
            commandline -f history-search-backward
        end
        commandline -f history-search-forward
        return
    end

    # We are not already in search mode.
    # If we are on the bottom line, start search mode,
    # otherwise move down
    set -l lineno (commandline -L)
    set -l line_count (count (commandline))

    switch $lineno
        case $line_count
            commandline -f history-search-forward

        case '*'
            commandline -f down-line
    end
end

bind \cp my_hist_back
bind \cn my_hist_forward
```
